# Project 9 -- [Drew] [Caldwell]

**TA Help:** N/A
    
**Collaboration:** N/A

## Question 1
<p>
Read in the ice cream products and reviews files into two separate data frames, as follows:

Notice that these two data frames have three columns in common, namely, brand, key, and ingredients. If we try to merge these two data frames by default, R will try to match data on all three columns, BUT the ingredients column has a different role in these two data frames. In the productsDF, the ingredients column has a list of the ice cream ingredients. In the ratingsDF, the ingredients column has values between 1 and 5, or an NA value.

For this reason, when we merge the information from the two tables, we only want to merge the data according to the brand and key values, as follows:

Notice that the productsDF has 8 columns, and the reviewsDF has 13 columns, and the newmergedDF has 19 columns, namely, all of the columns from both of the previous two data frames, including two separate ingredients columns, one from each data frame.

In this data frame, there are 978 rows that correspond to the name being "Chocolate Chip Cookie Dough", which has brand == "bj" and key == "16_bj". We can get this subset of the data as follows:

Notice that this new data frame called ChocolateChipCookieDoughDF has 978 rows.

What are the average number of stars for the 978 reviews in the data frame called ChocolateChipCookieDoughDF?
</p>

In [7]:
# Import library
library(data.table)
# read data
productsDF <- fread("/anvil/projects/tdm/data/icecream/combined/products.csv")
reviewsDF <- fread("/anvil/projects/tdm/data/icecream/combined/reviews.csv")

In [8]:
# Merge by brand and key
newmergedDF <- merge(productsDF, reviewsDF, by = c("brand", "key"))
# Subset for CC Cookie Dough
ChocolateChipCookieDoughDF <- subset(newmergedDF, (brand == "bj") & (key == "16_bj"))
print(nrow(ChocolateChipCookieDoughDF))

There are 978 rows as expected.

In [9]:
# Calc avg number of stars CC Cookie Dough
average_stars_ccd <- mean(ChocolateChipCookieDoughDF$stars, na.rm = TRUE)
print(average_stars_ccd)

[1] 4.560327


The average number of stars rated for the choclate chip cookie dough is 4.56

## Question 2

<p>
Starting with the newmergedDF, find the average number of stars for the reviews of ice cream with name == "Half Baked\302\256".

There are two characters that you will not see but they are there. They are encoded as "\302\256".

Print the average number of stars for the reviews of ice cream with name == "Half Baked\302\256".
</p>

In [17]:
# create new df that contains half baked reviews
half_baked_reviews <- newmergedDF[newmergedDF$name == "Half Baked\302\256", ]
# avg stars of half baked reviews
average_stars_half_baked <- mean(newmergedDF$stars[grepl("Half Baked", newmergedDF$name)], na.rm = TRUE)

In [18]:
# display avg
print(average_stars_half_baked)

[1] 4.699332


The average number of stars for the Half Baked is 4.69

## Question 3

<p>

In Project 2, we learned about the grep and grepl functions. Using either of these two functions, you can limit attention to ice cream products that have "Chocolate" in the title. (There are 49 such ice cream products.) Find the average number of stars for all 4831 of the reviews of these products that have "Chocolate" in the product name.

Print the average number of stars for all 4831 of the reviews of these products that have "Chocolate" in the product name.

</p>

In [12]:
# Avg for products w/ Choclate in name
chocolate_products <- newmergedDF[grepl("Chocolate", newmergedDF$name), ]
# get avg
average_stars_chocolate <- mean(chocolate_products$stars, na.rm = TRUE)
# output
print(average_stars_chocolate)

[1] 4.154833


The average rating on items with choclate in the flavor is 4.15

## Question 4
## It is necessary to have 2 cores in your Jupyter Lab session for Question 4 and Question 5.

<p>

Read in the information about airports, and also the flight data from 1990, into two separate data frames, as follows:

library(data.table)
airportsDF <- fread("/anvil/projects/tdm/data/flights/subset/airports.csv")
flightdataDF <- fread("/anvil/projects/tdm/data/flights/subset/1990.csv")

These two data frames do not have any columns in common! Nonetheless, the "iata" values from the airportsDF are the three-letter codes corresponding to airports, which are also found in the Origin and Dest columns in the flightdataDF. So when we merge the information from the two tables, if we want to study where the flights depart, then we only want to merge the data according to the iata value (from the airportsDF) merged with the Origin value (from the flightdataDF), as follows:

mybigDF <- merge(airportsDF, flightdataDF, by.x = "iata", by.y = "Origin")
Using this new data frame, find the average departure delay for all flights that have Origin airport in Indiana.
</p>

In [1]:
library(data.table)
airportsDF <- fread("/anvil/projects/tdm/data/flights/subset/airports.csv")
flightdataDF <- fread("/anvil/projects/tdm/data/flights/subset/1990.csv")

Warning message in fread("/anvil/projects/tdm/data/flights/subset/airports.csv"):
"Found and resolved improper quoting out-of-sample. First healed line 1253: <<"DBN","W. H. \"Bud\" Barron ","Dublin","GA","USA",32.56445806,-82.98525556>>. If the fields are not quoted (e.g. field separator does not appear within any field), try quote="" to avoid this warning."


In [2]:
# Clean the column names in airportsDF
setnames(airportsDF, gsub('"', '', names(airportsDF)))

# Merge by the iata
mybigDF <- merge(airportsDF, flightdataDF, by.x = "iata", by.y = "Origin")

# Avg departure delay for Indiana
indiana_airports <- airportsDF[airportsDF$state == "IN", ]
indiana_flights <- mybigDF[mybigDF$Origin %in% indiana_airports$iata, ]

In [21]:
print(head(indiana_airports))  # This will show all the Indiana airports and their details

Empty data.table (0 rows and 8 cols): iata,airport,city,state,country,lat...


In [19]:
print(nrow(indiana_flights))

[1] 0


In [3]:
# Calc the avg departure delay from indiana
average_departure_delay_indiana <- mean(indiana_flights$DepDelay, na.rm = TRUE)
print(average_departure_delay_indiana)

[1] NaN


It appears that there are no flight delays when leaving the indianapolis airport.

## Question 5

In [14]:
# Question 5: Average departure delay for flights from Houston, Texas
houston_airports <- airportsDF[airportsDF$city == "Houston" & airportsDF$state == "TX", ]
houston_flights <- mybigDF[mybigDF$Origin %in% houston_airports$iata, ]
head(houston_airports)

iata airport city state country lat long V8

In [16]:
#Avg departure houston
average_departure_delay_houston <- mean(houston_flights$DepDelay, na.rm = TRUE)

# Display avg departure delay
print(average_departure_delay_houston)

[1] NaN


In [17]:
# Check the houston_airports data frame
print(houston_airports)


Empty data.table (0 rows and 8 cols): iata,airport,city,state,country,lat...


It appears that there are no flights that leave from the Houston, TX airport.

## Pledge

By submitting this work I hereby pledge that this is my own, personal work. I've acknowledged in the designated place at the top of this file all sources that I used to complete said work, including but not limited to: online resources, books, and electronic communications. I've noted all collaboration with fellow students and/or TA's. I did not copy or plagiarize another's work.

> As a Boilermaker pursuing academic excellence, I pledge to be honest and true in all that I do. Accountable together – We are Purdue.